# A Case Study: The Effect of Gun Ownership on Gun-Homicide Rates


We consider the problem of estimating the effect of gun ownership on the homicide rate. For this purpose, we perform inference on $\beta$ in the following the partially linear model:
$$
Y_{j, t}=\beta D_{j,(t-1)}+g\left(X_{j, t}, \bar{X}_j, \bar{X}_t, X_{j, 0}, Y_{j, 0}, t\right)+\epsilon_{j, t}
$$
$Y_{j, t}$ is the log homicide rate in county $j$ at time $t. D_{j, t-1}$ is the log fraction of suicides committed with a firearm in county $j$ at time $t-1$, which we use as a proxy for gun ownership $G_{j, t}$, which is not observed. $X_{j, t}$ is a set of demographic and economic characteristics of county $j$ at time $t$. We use $\bar{X}_j$ to denote the within county average of $X_{j, t}$ and $\bar{X}_t$ to denote the within time period average of $X_{j, t} . X_{j, 0}$ and $Y_{j, 0}$ denote initial conditions in county $j$. We use $Z_{j, t}$ to denote the set of observed control variables $\left\{X_{j, t}, \bar{X}_j, \bar{X}_t, X_{j, 0}, Y_{j, 0}, t\right\}$, so that our model is

$$
 Y_{i,t} = \beta D_{i,(t-1)} + g(Z_{i,t}) + \epsilon_{i,t}.
$$

## Data

$Y_{j,t}$ is the log homicide rate in county $j$ at time $t$, $D_{j, t-1}$ is the log fraction of suicides committed with a firearm in county $j$ at time $t-1$, which we use as a proxy for gun ownership,  and  $Z_{j,t}$ is a set of demographic and economic characteristics of county $j$ at time $t$. Assuming the firearm suicide rate is a good proxy for gun ownership, the parameter $\beta$ is the effect of gun ownership on homicide rates, controlling for county-level demographic and economic characteristics.

The sample covers 195 large United States counties between the years 1980 through 1999, giving us 3900 observations.

In [1]:
# Import relevant packages
import pandas as pd
import numpy as np
import re
from sklearn.model_selection import cross_val_predict, KFold
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LassoCV, RidgeCV, ElasticNetCV, LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.neural_network import MLPRegressor
import warnings
import statsmodels.api as sm
import statsmodels.formula.api as smf
warnings.simplefilter('ignore')

np.random.seed(1234)

In [2]:
file = "https://raw.githubusercontent.com/CausalAIBook/MetricsMLNotebooks/main/data/gun_clean.csv"
data = pd.read_csv(file)
data.shape

(3900, 415)

## Preprocessing

To attempt to flexibly account for fixed heterogeneity across counties, common time factors, and deterministic time trends, we include county-level averages, time period averages, initial conditions, and the time index as additional control variables. This strategy is related to strategies for addressing latent sources of heterogeneity via conditioning.

We first reweight time and county variables as the data are population weighted.

In [3]:
# Note: These data are population weighted. Specifically,
# looking at the JBES replication files, they seem to be multiplied
# by sqrt((1/T sum_t population_{j,t})/100000). To get the
# unweighted variables need to divide by this number - which we can
# get from the time effects. We are mostly just going to use the weighted
# variables as inputs - except for time and county. We'll take
# cross-sectional and time series means of these weighted variables
# as well. Note that there is nothing wrong with this, but it does not
# reproduce a weighted regression in a setting where covariates may
# enter nonlinearly and flexibly.


# County FE
county_vars = data.filter(like='X_J')

# Time variables and population weights
# Pull out time variables
time_vars = data.filter(like='X_T')

# Use these to construct population weights
popweights = time_vars.sum(axis=1)

# Unweighted time variables
time_vars = time_vars.div(popweights, axis=0)

# For any columns with only zeros, drop them
time_vars = time_vars.loc[:, (time_vars != 0).any(axis=0)]

# Create time index
time_ind = (time_vars * np.arange(1, 21)).sum(axis=1)

Now we create initial conditions, county-level averages, and time period averages.

In [4]:
# Function to find variable names
def varlist(df=None, type=["numeric", "factor", "character"], pattern="", exclude=None):
    vars = []
    if any(t in type for t in ["numeric", "factor", "character"]):
        if "numeric" in type:
            vars += df.select_dtypes(include=["number"]).columns.tolist()
        if "factor" in type:
            vars += df.select_dtypes(include=["category"]).columns.tolist()
        if "character" in type:
            vars += df.select_dtypes(include=["object"]).columns.tolist()

    if exclude:
        vars = [var for var in vars if var not in exclude]

    if pattern:
        vars = [var for var in vars if re.search(pattern, var)]

    return vars


# Census control variables
census = []
census_var = ["^AGE", "^BN", "^BP", "^BZ", "^ED", "^EL", "^HI", "^HS",
              "^INC", "^LF", "^LN", "^PI", "^PO", "^PP", "^PV", "^SPR", "^VS"]

for pattern in census_var:
    census += varlist(df=data, pattern=pattern)

# Other control variables
X1 = ["logrobr", "logburg", "burg_missing", "robrate_missing"]
X2 = ["newblack", "newfhh", "newmove", "newdens", "newmal"]

# "Treatment" variable
d = "logfssl"

# Outcome variable
y = "logghomr"

# New DataFrame for time index
usedata = pd.DataFrame({'time.ind': time_ind})
usedata['weights'] = popweights

varlist_all = [y, d] + X1 + X2 + census
for var in varlist_all:
    usedata[var] = data[var]

# Construct initial conditions
varlist0 = [y] + X1 + X2 + census
for var in varlist0:
    usedata[f'{var}0'] = np.kron(usedata.loc[usedata['time.ind'] == 1, var], np.ones(20))

# County means
county_vars = pd.DataFrame(county_vars)
for var in varlist_all:
    usedata[f'{var}J'] = county_vars.dot(np.linalg.pinv(county_vars).dot(usedata[var]))

# Time means
time_vars = pd.DataFrame(time_vars)
for var in varlist_all:
    usedata[f'{var}T'] = time_vars.dot(np.linalg.pinv(time_vars).dot(usedata[var]))


# Estimation

## Baseline OLS Estimates

After preprocessing the data, as a baseline model, we first look at simple regression of $Y_{j,t}$ on $D_{j,t-1}$ without controls in the full data set.

In [5]:
# Baseline OLS
X = sm.add_constant(usedata['logfssl'])
y = usedata['logghomr']
lm0 = sm.OLS(y, X).fit(cov_type='HC3')
vc0 = lm0.cov_params()
coef = lm0.params['logfssl']
std_err = np.sqrt(vc0.loc['logfssl', 'logfssl'])
print("Baseline OLS:", coef, "(", std_err, ")")

Baseline OLS: 0.30202337690461367 ( 0.012565715795207516 )


The point estimate is $0.302$ with the confidence interval ranging from 0.277 to 0.327. This
suggests that increases in gun ownership rates are related to gun homicide rates - if gun ownership increases by 1% then the predicted gun homicide rate goes up by 0.3%, without controlling for counties' characteristics.



Next we estimate with the baseline set of controls.

In [6]:
# Regression on baseline controls
varlist = [d] + X1 + X2 + census
X = sm.add_constant(usedata[varlist])
y = usedata['logghomr']
lmC = sm.OLS(y, X).fit(cov_type='HC3')
vcC = lmC.cov_params()
coef = lmC.params['logfssl']
std_err = np.sqrt(vcC.loc['logfssl', 'logfssl'])
print("OLS with Controls:", coef, "(", std_err, ")")

OLS with Controls: 0.3102764695463243 ( 0.0728034338946108 )


<!-- Since our goal is to estimate the effect of gun ownership after controlling for a rich set county characteristics, we next include time and space averages. -->

We can also run our regression with time and space averages as controls.

In [7]:
# Regression on time and cross-sectional averages
varlistX = X1 + X2 + census
varlistMeans = [d] + X1 + X2 + census

# Adding county and time means
for var in varlistX:
    varlistMeans.append(var + 'J')
    varlistMeans.append(var + 'T')

X = sm.add_constant(usedata[varlistMeans])
y = usedata['logghomr']
lmM = sm.OLS(y, X).fit(cov_type='HC3')
vcM = lmM.cov_params()
coef = lmM.params['logfssl']
std_err = np.sqrt(vcM.loc['logfssl', 'logfssl'])
print("OLS with Averages:", coef, "(", std_err, ")")

OLS with Averages: 0.16784092104187234 ( 3.526533013652346 )


Since our goal is to estimate the effect of gun ownership after controlling for a rich set county characteristics, we now include all controls.

In [8]:
# Regression on all controls

X = sm.add_constant(usedata.drop(columns=['logghomr']))
y = usedata['logghomr']
lmA = sm.OLS(y, X).fit(cov_type='HC3')
vcA = lmA.cov_params()
coef = lmA.params['logfssl']
std_err = np.sqrt(vcA.loc['logfssl', 'logfssl'])
print("OLS All:", coef, "(", std_err, ")")

OLS All: 0.17931589905305267 ( 0.06762548941017878 )


After controlling for a rich set of characteristics, the point estimate of gun ownership attenuates to 0.179.

***NB***: In the background, `statsmodels` is dropping variables based on collinearity diagnostics. These depend on system linear algebra routines and can lead to large differences in high-dimensional or other ill-conditioned settings when using otherwise identical code across languages and/or machines.

Now we turn to our double machine learning framework, employing linear and flexible estimation methods with cross-fitting.

## DML Estimates

We perform inference on $\beta$ in the following the partially linear model:
 $$
Y_{j, t}=\beta D_{j,(t-1)}+g(Z_{j,t})+\epsilon_{j, t}.
$$
In the first stage, using cross-fitting, we employ modern regression methods to build estimators $\hat \ell(Z_{j,t})$ and $\hat m(Z_{j,t})$, where
- $\ell(Z_{j,t}):=E(Y_{j,t}|Z_{j,t})$
- $m(Z_{j,t}):=E(D_{j,t}|Z_{j,t})$

Using these, we obtain the estimates of the residualized quantities
- $\tilde Y_{j,t} = Y_{j,t}- E(Y_{j,t}|Z_{j,t})$
- $\tilde D_{j,t}= D_{j,t}- E(D_{j,t}|Z_{j,t})$

Using these residualized quantities, we note our model can be written as
$$
\tilde Y_{j,t} = \beta \tilde D_{j,t} + \epsilon_{j,t}, \quad E (\epsilon_{j,t} |\tilde D_{j,t}) =0.
$$
In the final stage, using ordinary least squares of $\tilde Y_{j,t}$ on $\tilde D_{j,t}$, we obtain the
estimate of $\beta$.

In the following, we consider 10 different methods for the first-stage models for $\ell(\cdot)$ and $m(\cdot)$ covering linear, penalized linear, and flexible methods. We also report the first-stage RMSE scores for estimating $Y$ and $D$.

In [9]:
def dml(X, D, y, modely, modeld, *, nfolds, classifier=False):
    '''
    DML for the Partially Linear Model setting with cross-fitting

    Input
    -----
    X: the controls
    D: the treatment
    y: the outcome
    modely: the ML model for predicting the outcome y
    modeld: the ML model for predicting the treatment D
    nfolds: the number of folds in cross-fitting
    classifier: bool, whether the modeld is a classifier or a regressor

    time: array of time indices, eg [0,1,...,T-1,0,1,...,T-1,...,0,1,...,T-1]
    clu: array of cluster indices, eg [1073, 1073, 1073, ..., 5055, 5055, 5055, 5055]
    cluster: bool, whether to use clustered standard errors

    Output
    ------
    point: the point estimate of the treatment effect of D on y
    stderr: the standard error of the treatment effect
    yhat: the cross-fitted predictions for the outcome y
    Dhat: the cross-fitted predictions for the treatment D
    resy: the outcome residuals
    resD: the treatment residuals
    epsilon: the final residual-on-residual OLS regression residual
    '''
    cv = KFold(n_splits=nfolds, shuffle=True, random_state=123)  # shuffled k-folds
    yhat = cross_val_predict(modely, X, y, cv=cv, n_jobs=-1)  # out-of-fold predictions for y
    # out-of-fold predictions for D
    # use predict or predict_proba dependent on classifier or regressor for D
    if classifier:
        Dhat = cross_val_predict(modeld, X, D, cv=cv, method='predict_proba', n_jobs=-1)[:, 1]
    else:
        Dhat = cross_val_predict(modeld, X, D, cv=cv, n_jobs=-1)
    # calculate outcome and treatment residuals
    resy = y - yhat
    resD = D - Dhat

    dml_data = pd.concat([pd.Series(resy, name='resy'), pd.Series(resD, name='resD')], axis=1)
    ols_mod = smf.ols(formula='resy ~ 1 + resD', data=dml_data).fit()

    point = ols_mod.params[1]
    stderr = ols_mod.bse[1]
    epsilon = ols_mod.resid

    return point, stderr, yhat, Dhat, resy, resD, epsilon

In [10]:
def summary(point, stderr, yhat, Dhat, resy, resD, epsilon, X, D, y, *, name):
    '''
    Convenience summary function that takes the results of the DML function
    and summarizes several estimation quantities and performance metrics.
    '''
    return pd.DataFrame({'estimate': point,  # point estimate
                         'stderr': stderr,  # standard error
                         'lower': point - 1.96 * stderr,  # lower end of 95% confidence interval
                         'upper': point + 1.96 * stderr,  # upper end of 95% confidence interval
                         'rmse y': np.sqrt(np.mean(resy**2)),  # RMSE of model that predicts outcome y
                         'rmse D': np.sqrt(np.mean(resD**2))  # RMSE of model that predicts treatment D
                         }, index=[name])

In [11]:
# OLS No Controls
Y = usedata['logghomr']
D = usedata['logfssl']
Z = pd.DataFrame({"Const": np.ones(len(Y))})  # regression on constant

modely = make_pipeline(StandardScaler(), LinearRegression())
modeld = make_pipeline(StandardScaler(), LinearRegression())

# Run DML model with nfolds folds of cross-fitting
result_OLS = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
table = summary(*result_OLS, Z, D, y, name='OLS No Controls')
table

estimate    stderr     lower     upper    rmse y    rmse D
OLS No Controls   0.30196  0.013676  0.275155  0.328764  1.096108  1.210306

In [12]:
# Basic Controls
basic_controls = [d] + X1 + X2 + census
Z = usedata[basic_controls].drop(columns=['logfssl'])

modely = make_pipeline(StandardScaler(), LinearRegression())
modeld = make_pipeline(StandardScaler(), LinearRegression())

# Run DML model with nfolds folds of cross-fitting
result_basic = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
table = pd.concat([table, summary(*result_basic, Z, D, y, name='OLS Basic Controls')])
table

estimate    stderr     lower     upper    rmse y    rmse D
OLS No Controls     0.301960  0.013676  0.275155  0.328764  1.096108  1.210306
OLS Basic Controls  0.352414  0.061963  0.230967  0.473862  0.499377  0.128553

In [13]:
# All Controls
# Regression on time and cross-sectional averages
varlistX = X1 + X2 + census
varlistMeans = [d] + X1 + X2 + census

# Adding county and time means
for var in varlistX:
    varlistMeans.append(var + 'J')
    varlistMeans.append(var + 'T')

Z = sm.add_constant(usedata[varlistMeans].drop(columns=['logfssl']))

modely = make_pipeline(StandardScaler(), LinearRegression())
modeld = make_pipeline(StandardScaler(), LinearRegression())

# Run DML model with nfolds folds of cross-fitting
result_all = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
table = pd.concat([table, summary(*result_all, Z, D, y, name='OLS All Controls')])
table

estimate    stderr     lower     upper    rmse y    rmse D
OLS No Controls     0.301960  0.013676  0.275155  0.328764  1.096108  1.210306
OLS Basic Controls  0.352414  0.061963  0.230967  0.473862  0.499377  0.128553
OLS All Controls    0.111572  0.053659  0.006400  0.216743  0.425974  0.127081

In [14]:
# Now lets do Cross-validated Lasso, Ridge, ENet
cv = KFold(n_splits=10, shuffle=True, random_state=123)  # shuffled k-folds

In [15]:
# Define LassoCV models with n_splits folds of cross-validation
modely = make_pipeline(StandardScaler(), LassoCV(cv=cv))
modeld = make_pipeline(StandardScaler(), LassoCV(cv=cv))

# Run DML model with nfolds folds of cross-fitting
result_LassoCV = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
table = pd.concat([table, summary(*result_LassoCV, Z, D, y, name='LassoCV')])
table

estimate    stderr     lower     upper    rmse y    rmse D
OLS No Controls     0.301960  0.013676  0.275155  0.328764  1.096108  1.210306
OLS Basic Controls  0.352414  0.061963  0.230967  0.473862  0.499377  0.128553
OLS All Controls    0.111572  0.053659  0.006400  0.216743  0.425974  0.127081
LassoCV             0.517538  0.057333  0.405165  0.629911  0.531980  0.147088

In [ ]:
# Define RidgeCV models with n_splits folds of cross-validation
modely = make_pipeline(StandardScaler(), RidgeCV(cv=cv))
modeld = make_pipeline(StandardScaler(), RidgeCV(cv=cv))

# Run DML model with nfolds folds of cross-fitting
result_RidgeCV = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
table = pd.concat([table, summary(*result_RidgeCV, Z, D, y, name='RidgeCV')])
table

In [ ]:
# Define ElasticNetCV models with n_splits folds of cross-validation
modely = make_pipeline(StandardScaler(), ElasticNetCV(l1_ratio=0.5, cv=cv))
modeld = make_pipeline(StandardScaler(), ElasticNetCV(l1_ratio=0.5, cv=cv))

# Run DML model with nfolds folds of cross-fitting
result_ENetCV = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
table = pd.concat([table, summary(*result_ENetCV, Z, D, y, name='ENetCV')])
table

In [ ]:
# DML with Random Forests. RFs don't require scaling but we do it for consistency
modely = make_pipeline(StandardScaler(), RandomForestRegressor(n_estimators=100, min_samples_leaf=5, random_state=123))
modeld = make_pipeline(StandardScaler(), RandomForestRegressor(n_estimators=100, min_samples_leaf=5, random_state=123))

# Run DML model with nfolds folds of cross-fitting (computationally intensive)
result_RF = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
table = pd.concat([table, summary(*result_RF, Z, D, y, name='RF')])
table

In [ ]:
# DML with Boosted Trees
modely = make_pipeline(StandardScaler(), GradientBoostingRegressor(max_depth=4, n_iter_no_change=5))
modeld = make_pipeline(StandardScaler(), GradientBoostingRegressor(max_depth=4, n_iter_no_change=5))

# Run DML model with nfolds folds of cross-fitting (computationally intensive)
result_BT = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
table = pd.concat([table, summary(*result_BT, Z, D, y, name='Boosted Trees')])
table

In [ ]:
# DML with NNs
modely = make_pipeline(StandardScaler(),
                       MLPRegressor(hidden_layer_sizes=(50, 50, 50, 50),
                                    activation='relu',
                                    solver='adam',
                                    alpha=0.0001,
                                    batch_size=200,
                                    learning_rate='constant',
                                    learning_rate_init=0.001,
                                    max_iter=200,
                                    shuffle=True,
                                    random_state=None,
                                    tol=1e-4,
                                    verbose=False,
                                    warm_start=False,
                                    momentum=0.9,
                                    nesterovs_momentum=True,
                                    early_stopping=True,
                                    validation_fraction=0.2,
                                    beta_1=0.9,
                                    beta_2=0.999,
                                    epsilon=1e-08,
                                    n_iter_no_change=10)
                       )
modeld = make_pipeline(StandardScaler(),
                       MLPRegressor(hidden_layer_sizes=(50, 50, 50, 50),
                                    activation='relu',
                                    solver='adam',
                                    alpha=0.0001,
                                    batch_size=200,
                                    learning_rate='constant',
                                    learning_rate_init=0.001,
                                    max_iter=200,
                                    shuffle=True,
                                    random_state=None,
                                    tol=1e-4,
                                    verbose=False,
                                    warm_start=False,
                                    momentum=0.9,
                                    nesterovs_momentum=True,
                                    early_stopping=True,
                                    validation_fraction=0.2,
                                    beta_1=0.9,
                                    beta_2=0.999,
                                    epsilon=1e-08,
                                    n_iter_no_change=10)
                       )

# Run DML model with nfolds folds of cross-fitting
result_NN = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
table = pd.concat([table, summary(*result_NN, Z, D, y, name='NN (Early Stopping)')])
table

In [ ]:
rmses = table.iloc[:, -2:]
rmses

In [ ]:
print("Lowest RMSE y: ", rmses.iloc[:, 0].idxmin())
print("Lowest RMSE D: ", rmses.iloc[:, 1].idxmin())

In [ ]:
# DML with best model, which is RF
modely = make_pipeline(StandardScaler(), RandomForestRegressor(n_estimators=100, min_samples_leaf=5, random_state=123))
modeld = make_pipeline(StandardScaler(), RandomForestRegressor(n_estimators=100, min_samples_leaf=5, random_state=123))

# Run DML model with nfolds folds of cross-fitting (computationally intensive)
result_best = dml(Z, D, Y, modely, modeld, nfolds=5, classifier=False)
table = pd.concat([table, summary(*result_best, Z, D, y, name='Best')])
table

In [ ]:
# Least Squares Model Average
yhat_all = pd.concat([
    pd.Series(result_OLS[2]),
    pd.Series(result_basic[2]),
    pd.Series(result_all[2]),
    pd.Series(result_LassoCV[2]),
    pd.Series(result_RidgeCV[2]),
    pd.Series(result_ENetCV[2]),
    pd.Series(result_RF[2]),
    pd.Series(result_BT[2]),
    pd.Series(result_NN[2])
], axis=1)

Dhat_all = pd.concat([
    pd.Series(result_OLS[3]),
    pd.Series(result_basic[3]),
    pd.Series(result_all[3]),
    pd.Series(result_LassoCV[3]),
    pd.Series(result_RidgeCV[3]),
    pd.Series(result_ENetCV[3]),
    pd.Series(result_RF[3]),
    pd.Series(result_BT[3]),
    pd.Series(result_NN[3])
], axis=1)

ma_y = sm.OLS(usedata['logghomr'], yhat_all).fit()
ma_d = sm.OLS(usedata['logfssl'], Dhat_all).fit()

weights_y = ma_y.params
weights_d = ma_d.params

lm_k = sm.OLS(ma_y.resid, ma_d.resid).fit(cov_type='HC3')
lsma = pd.Series({"estimate": lm_k.params[0], "stderr": lm_k.bse[0]},
                 name="Least Squares Model Average").to_frame().T

final_table = table.iloc[:, :2]
final_table = pd.concat([final_table, lsma], axis=0)
final_table